### PR TITLE
Forms: prevent PHP fatals

### PR DESCRIPTION
**Pull request by monsieur-z**
## Proposed changes:

This PR updates the `forms` package to prevent PHP fatals that have been raised.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p1722410167616409-slack-C034JEXD1RD

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

Review the code. Let's check the form functionalities still work:
- Spin up a test site with this branch
- Create a new post
- Add a Contact Form block
- Save the post and open the preview
- Fill the form and submit it
- Notice it works as expected

Optionally, test that the Google Drive export works.
- Visit `/wp-admin/admin.php?page=jetpack-forms#/responses`
- Select the `Inbox` view
Screenshot 2024-07-31 at 10 42 33 AM

- Verify that at least one item is present in the `Inbox` tab
Screenshot 2024-07-31 at 10 51 34 AM

- Click the `Export` button and  follow the connection process
Screenshot 2024-07-31 at 10 15 11 AM

- A new tab with a Google Spreadsheet should open

### CODE REVIEW

File: projects/packages/forms/src/contact-form/class-contact-form.php [hunk 1]
```diff
@@ -310,7 +310,7 @@ public static function parse( $attributes, $content ) {
 				"</h4>\n\n";
 
 			// Don't show the feedback details unless the nonce matches
-			if ( $feedback_id && isset( $_GET['_wpnonce'] ) && wp_verify_nonce( stripslashes( $_GET['_wpnonce'] ), "contact-form-sent-{$feedback_id}" ) ) { // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			if ( $feedback_id && isset( $_GET['_wpnonce'] ) && is_string( $_GET['_wpnonce'] ) && wp_verify_nonce( stripslashes( $_GET['_wpnonce'] ), "contact-form-sent-{$feedback_id}" ) ) { // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
Comment (monsieur-z) [inline projects/packages/forms/src/contact-form/class-contact-form.php]: Prevents following error: `Uncaught TypeError: stripslashes(): Argument #1 ($string) must be of type string, array given`

I haven't been able to test this path, but the change looks straightforward enough.

File: projects/packages/forms/src/contact-form/class-admin.php [hunk 2]
```diff
@@ -190,7 +190,7 @@ public function export_to_gdrive() {
 		$grunion     = Contact_Form_Plugin::init();
 		$export_data = $grunion->get_feedback_entries_from_post();
 
-		$fields    = array_keys( $export_data );
+		$fields    = is_array( $export_data ) ? array_keys( $export_data ) : array();
```
Comment (monsieur-z) [inline projects/packages/forms/src/contact-form/class-admin.php:193]: Prevents following error: `Uncaught TypeError: array_keys(): Argument #1 ($array) must be of type array, null given`

File: projects/packages/forms/src/contact-form/class-contact-form.php [hunk 3]
```diff
@@ -1050,35 +1050,44 @@ public function process_submission() {
...
-			$field          = $this->fields[ $field_ids['name'] ];
-			$comment_author = Contact_Form_Plugin::strip_tags(
-				stripslashes(
-					/** This filter is already documented in core/wp-includes/comment-functions.php */
-					apply_filters( 'pre_comment_author_name', addslashes( $field->value ) )
-				)
-			);
+			$field = $this->fields[ $field_ids['name'] ];
+
+			if ( is_string( $field->value ) ) {
```
Comment (monsieur-z) [inline projects/packages/forms/src/contact-form/class-contact-form.php:1059]: This guard and the next ones prevent the following error: `Uncaught TypeError: addslashes(): Argument #1 ($string) must be of type string, array given`

I haven't been able to find a case where `$field->value` was something other than a string, so I don't know what a useful fallback could be.

**Review by coder-karen — COMMENTED** [review]
I tested on self-hosted and also Simple (copying changes over to the relevant files). I couldn't reproduce any errors 'naturally' either, and form submission plus connection to Google all worked smoothly 🥳 

I did try to directly modify values in the codebase - I set `$field->value` to `[ "foo" => "bar" ]` when setting the name within the existing code, and I can replicate the critical error.

With the changes, if I set that array as the` $field->value` value, I got a different fatal error -` PHP Fatal error:  Uncaught TypeError: explode(): Argument #2 ($string) must be of type string, array given in /usr/local/src/jetpack-monorepo/projects/plugins/crm/includes/ZeroBSCRM.Jetpack.php:131)`. The line in question in the CRM codebase is `$name  = explode( ' ', $field_data->value, 2 );`

I don't know if this is replicable in real-time somehow (is another plugin modifying the field value before it's getting to the contact form?), but just in case perhaps we should ensure if the `$field->value` in those changed cases is not a string, that is converted to a blank string or something similar?

**Comment by monsieur-z** [comment]
Thank you @coder-karen!

Do you mean something like what I did in [this commit?](https://github.com/Automattic/jetpack/pull/38647/commits/8938ce43f6b83e781e47a14f77e8026b2cb89d99)

**Comment by coder-karen** [comment]
Ah interesting, actually I was thinking outside the `strip_tags function` and outside the previously added `is_string` check. The issue in my testing was if `$field->value` was an array. If so, it won't go through that `strip_tags` / `stripslashes` function, therefore the previous error won't be generated and code would continue to run.

With the change you've just made, `$field->value` would retain the array value - I also get the same CRM related critical error with that if I set `$field->value = [ "foo" => "bar" ];` just above the `$comment_author = ` line. 
So I'm wondering if with the previous code where we check if the value is a string, we have an else block afterwards which changes `$field->value`  to an empty string if it isn't one.

File: projects/packages/forms/src/contact-form/class-contact-form.php (hunk 1 above)
Comment (jeherve) [reply projects/packages/forms/src/contact-form/class-contact-form.php]: Should we take that opportunity to clean up a bit how we check for that nonce? Something like 

```suggestion
			if (
				$feedback_id
				&& isset( $_GET['_wpnonce'] )
				&& wp_verify_nonce( sanitize_key( wp_unslash( $_GET['_wpnonce'] ) ), "contact-form-sent-{$feedback_id}" )
			) {
```

**Comment by monsieur-z** [comment]
I understand, now. Thanks @coder-karen!

The latest commit seems to have fixed the CRM error.

**Review by coder-karen — APPROVED** [review]
Tests well for me, can't replicate any errors now, thanks!

**Pull request merged**